### PR TITLE
abstracted classes and methods for transformer contextual lenses

### PR DESCRIPTION
**Pull request by amirshane**

### CODE REVIEW

File: transformer_contextual_lenses.py [hunk 1]
```diff
@@ -0,0 +1,706 @@
...
+"""# Transformer model
+Code source: https://github.com/google/flax/blob/master/examples/lm1b/models.py
+"""
+
+def shift_right(x):
+  """Shift the input to the right by padding on axis 1."""
+  pad_widths = [(0, 0)] * len(x.shape)
+  pad_widths[1] = (1, 0)  # Padding on axis=1
+  padded = jnp.pad(
+      x, pad_widths, mode='constant', constant_values=x.dtype.type(0))
```
Comment (dmrd) [inline transformer_contextual_lenses.py:62]: Not a bug in current usage, but something to be aware of: Both PAD and BOS (beginning of sentence) are 0 in the LM1B example (and therefore this fork). This is prone to bugs down the line.

File: transformer_contextual_lenses.py [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+
+"""# Transformer lenses
+Based on: https://arxiv.org/pdf/2002.08866.pdf
+
+### Lens 1: Pooling
+
+Mean pooling
+"""
+
+class TransformerMeanPool(nn.Module):
```
Comment (dmrd) [inline transformer_contextual_lenses.py:357]: Instead of making each of these a separate class, how about adding a `reduce_fn` argument to TransformerLM. Can then add
```
if reduce_fn is not None:
  x = reduce_fn(x)
```

Furthermore, can make reduce_fn act on (x, padding_mask), to allow it to ignore padding positions.


```
if reduce_fn is not None:
  x = reduce_fn(x, padding_mask)
```

You can then create the pooling models with:

TransformerMeanPool = TransformerLM.partial(reduce_fn=mean_reduce_fn)

etc
